### PR TITLE
Add local gradient aggregation to tf.keras.

### PR DESCRIPTION
**Pull request by bfrieds**

### CODE REVIEW

File: horovod/_keras/__init__.py [hunk 1]
```diff
@@ -30,6 +30,18 @@ def __init__(self, name, device_dense, device_sparse, compression, sparse_as_den
...
             self._sparse_as_dense = sparse_as_dense
             self._get_gradients_used = False
+
+            # How often are parameters synchronized
+            self._aggregation_frequency = aggregation_frequency
+            assert self._aggregation_frequency > 0
+
+            # This is going to be N data structure holding the per-GPU aggregated gradient updates
+            # for parameter updates. N is the number of parameters.
+            self.gpu_shadow_vars = []
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: The per-GPU comment is unnecessary here as each GPU operates independently.

File: horovod/_keras/__init__.py [hunk 2]
```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
...
             """
-            self._get_gradients_used = True
-            gradients = super(self.__class__, self).get_gradients(loss, params)
-            if hvd.size() > 1:
-                averaged_gradients = []
-                with tf.name_scope(self._name + "_Allreduce"):
-                    for grad in gradients:
-                        if grad is not None:
+            def init_aggregation_vars():
+                v = tf.get_collection('aggregation_variables')
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: collections actually have some use !!!

File: horovod/_keras/__init__.py [hunk 3]
```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
...
+                with tf.variable_scope("aggregation_variables"):
+                    self.counter = tf.get_variable(
+                        "aggregation_counter", shape=(), dtype=tf.int32,
+                        trainable=False, initializer=tf.zeros_initializer())
+                    if self._aggregation_frequency > 1:
+                        for idx, grad in enumerate(self.grads):
                             if self._sparse_as_dense and \
                                     isinstance(grad, tf.IndexedSlices):
                                 grad = tf.convert_to_tensor(grad)
+                            grad_aggregation_variable_name = str(idx)
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: Should we put a check in here that `grad` is not `isinstance(grad, tf.IndexedSlices)`? Haven't dove too deep on this but it seems to me that our current gradient aggregation techniques would not support it.

File: horovod/_keras/__init__.py [hunk 4]
```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
...
+                    with tf.control_dependencies([g.op for g in averaged_gradients]):
+                        reset_op = self.counter.assign(
+                            tf.constant(0), use_locking=True)
+                    with tf.control_dependencies([reset_op]):
+                        return [tf.identity(g) for g in averaged_gradients]
+
+            self._get_gradients_used = True
+            self.grads = super(
+                self.__class__, self).get_gradients(loss, params)
+            init_aggregation_vars()
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: Why doesn't this call re-initialize the variables every time ?

File: horovod/_keras/__init__.py [hunk 5]
```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
...
+                    aggregation_read_ops = tf.group(
+                        *aggregation_read_ops_list)
+                else:
+                    aggregated_grads = self.grads
+                    aggregation_read_ops = tf.no_op()
+
+                with tf.control_dependencies([aggregation_read_ops]):
+                    averaged_gradients = []
+                    for idx, grad in enumerate(aggregated_grads):
+                        if grad is not None:
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: We want to average our locally aggregated gradients by `_aggregation_frequency` prior to calling `hvd.allreduce()`. Or after is fine also. Possibly this could replace the useless `tf.identity()` function we have in the return statement of this function.

File: horovod/_keras/__init__.py [hunk 6]
```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
...
-                  raise Exception('`apply_gradients()` was called without a call to '
-                                  '`get_gradients()`. If you\'re using TensorFlow 2.0, '
-                                  'please specify `experimental_run_tf_function=False` in '
-                                  '`compile()`.')
-            return super(self.__class__, self).apply_gradients(*args, **kwargs)
+                raise Exception('`apply_gradients()` was called without a call to '
+                                '`get_gradients()`. If you\'re using TensorFlow 2.0, '
+                                'please specify `experimental_run_tf_function=False` in '
+                                '`compile()`.')
+            flattended_args0 = [item for tup in args[0] for item in tup]
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: Maybe add a comment for why we need this?

File: horovod/_keras/__init__.py (hunk 5 above)
Comment (bfrieds) [reply horovod/_keras/__init__.py]: yeah I like this idea of replacing that `tf.identity()` with the div by `self._aggregation_frequency`

File: horovod/_keras/__init__.py (hunk 4 above)
Comment (bfrieds) [reply horovod/_keras/__init__.py]: Here's the best explanation I can come up with. `init_aggregation_vars` has no reference to it later in the function. So there's no dependency placed on it by anything else that sets up the graph. Everything else can be traced back from the return. So it will get called once when initializing the graph, but won't be called again ever again. This was verified by some prints.

File: horovod/_keras/__init__.py (hunk 3 above)
Comment (bfrieds) [reply horovod/_keras/__init__.py]: just fail in that case?

Comment (aaron276h) [reply horovod/_keras/__init__.py]: Yeah just print  if `aggregation_frequency > 1` we don't support sparse communication. On second look we might need to do this before the `sparse_as_dense` check also.

Comment (aaron276h) [reply horovod/_keras/__init__.py]: Can we move this up above ` if self._sparse_as_dense and...`
